Objects/odictobject.c

```diff
@@ -1356,28 +1356,17 @@ static PyGetSetDef odict_getset[] = {
 static void
 odict_dealloc(PyODictObject *self)
 {
-    PyThreadState *tstate = _PyThreadState_GET();
-
     PyObject_GC_UnTrack(self);
-    Py_TRASHCAN_SAFE_BEGIN(self)
+    Py_TRASHCAN_BEGIN(self, odict_dealloc)
 
     Py_XDECREF(self->od_inst_dict);
     if (self->od_weakreflist != NULL)
         PyObject_ClearWeakRefs((PyObject *)self);
 
     _odict_clear_nodes(self);
-
-    /* Call the base tp_dealloc().  Since it too uses the trashcan mechanism,
-     * temporarily decrement trash_delete_nesting to prevent triggering it
-     * and putting the partially deallocated object on the trashcan's
-     * to-be-deleted-later list.
-     */
-    --tstate->trash_delete_nesting;
-    assert(_tstate->trash_delete_nesting < PyTrash_UNWIND_LEVEL);
     PyDict_Type.tp_dealloc((PyObject *)self);
```

Comment:
By the way, why is it not needed to retrack the object here?

Reply:
> By the way, why is it not needed to retrack the object here?

Because there is no general requirement to retrack. A typical `tp_dealloc` implementation uses `PyObject_GC_UnTrack()` which is a no-op if the object is already untracked. Retracking is only needed if the `tp_dealloc` of the base class uses unsafe GC macros that might crash on non-GC-tracked objects.

Conclusion: in the case where we don't know the base class, we need to retrack. If we know that the base class is fine, there is no need.

Reply:
This sounds like something to carve into a stone somewhere.

Reply:
It's already documented in `Objects/typeobject.c`:
```
       Q. Why do we GC-untrack before the trashcan and then immediately
          GC-track again afterward?

       A. In the case that the base class is GC-aware, the base class
          probably GC-untracks the object.      If it does that using the
          UNTRACK macro, this will crash when the object is already
          untracked.  Because we don't know what the base class does, the
          only safe thing is to make sure the object is tracked when we
          call the base class dealloc.  But...  The trashcan begin macro
          requires that the object is *untracked* before it is called.  So
          the dance becomes:

         GC untrack
         trashcan begin
         GC track
```
As far as I can tell, this is the only place in CPython where we do retrack before calling the base `tp_dealloc`.